mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd

```diff
@@ -0,0 +1,191 @@
+```
+
+# Forward Hotstarting
+
+Forward hotstarting is currently only supported by XGBoost.
+However, we have observed that hotstarting only provides a speed advantage for very large datasets and models with more than 5000 boosting rounds.
+The reason is that copying the models from the main process to the workers is a major bottleneck.
+The parallelization package future copies the models sequentially to the workers.
+Consequently, it takes a long time until the last worker can even start.
+Moreover, copying itself consumes a lot of time, and copying the model back from the worker blocks the main process again.
+During the development process, we overestimated the speed benefits of hotstarting and underestimated the overhead of parallelization.
+We can therefore only advise against using forward hotstarting during tuning.
+It is much more efficient to use the internal early-stopping mechanism of XGBoost.
+This eliminates the need to copy models to the worker.
+See the [gallery post](/gallery/optimization/2022-11-04-early-stopping-with-xgboost) on early stopping for an example.
+We might fix forward hotstarting in the future with workers that read hotstart models from a database in parallel.
+However, the need for this seems to be rather small.
+Models that can be trained iteratively usually have a very efficient internal early-stopping mechanism.
+
+# Manual Hotstarting
+
+Nevertheless, forward hotstarting can be useful without parallelization.
+If you have an already trained model and want to add more boosting iteration to it.
+In this example, the `learner_5000` is the already trained model.
+We create a new learner with the same hyperparameters but double the number of boosting iteration.
+To activate hotstarting, we create a `HotstartStack` and copy it to the `$hotstart_stack` slot of the new learner.
+
+```{r}
+#| eval: false
+
+library(mlr3verse)
+task = tsk("spam")
+
+learner_5000 = lrn("classif.xgboost", nrounds = 5000, eta = 0.1)
+learner_5000$train(task)
+
+learner_10000 = lrn("classif.xgboost", nrounds = 10000, eta = 0.1)
+learner_10000$hotstart_stack = HotstartStack$new(learner_5000)
+learner_10000$train(task)
+```
+
+```{r}
+#| echo: false
+
+learner_5000 = readRDS("learner_5000.rds")
+learner_10000 = readRDS("learner_10000.rds")
+```
+
+Training the initial model took `r learner_5000$state$train_time` seconds.
+
+```{r}
+learner_5000$state$train_time
+```
+
+Adding 5000 boosting rounds took `r learner_10000$state$train_time - learner_5000$state$train_time` seconds.
+
+```{r}
+learner_10000$state$train_time - learner_5000$state$train_time
+```
+
+Training the model from the beginning would have taken about two minutes.
+This means, without parallelization, we get the expected speed advantage.
+
+# Conclusion
+
+Backward hotstarting significantly reduces the time to tune a random forest.
+Whereas forward hotstarting is less useful due to the parallelization overhead and the very efficient internal early stopping mechanisms.
```

Comment:
```suggestion
One has to be careful, however, when using forward hotstarting during tuning because of the high parallelization overhead that arises from copying the models between the processes. 
If a model has an internal early stopping implementation, it should usually be relied upon instead of using the mlr3 hotstarting mechanism.
```